File: q1544/q1544.py
class Solution:
    def makeGood(self, s: str) -> str:
        stack = []
        for c in s: 
            # Check if the incoming character will result in a bad pair
            if len(stack) != 0 and (c != stack[-1] and c.lower() == stack[-1].lower()): 
                # Do not add incoming char, and remove the previous char too
                stack.pop()
            else: 
                stack.append(c)
        
        return "".join(stack)

File: q1544/q1544.java
class Solution {
    public String makeGood(String s) {
        ArrayDeque<Character> stack = new ArrayDeque<>();

        for (char c : s.toCharArray()) {
            // Check if the incoming character will result in a bad pair
            if (!stack.isEmpty() &&
                    (c != stack.peek() && Character.toLowerCase(c) == Character.toLowerCase(stack.peek()))) {
                // Do not add the incoming character, and remove the previous one too
                stack.pop();
            } else {
                stack.push(c);
            }
        }

        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        return sb.reverse().toString();
    }
}